src/main/java/org/truffleruby/language/methods/LookupMethodNode.java

```diff
@@ -128,16 +128,16 @@ protected InternalMethod lookupMethodUncached(
             throw Utils.unsupportedOperation("method lookup not supported on foreign objects");
         }
 
+        final DeclarationContext declarationContext = RubyArguments.tryGetDeclarationContext(frame);
         final InternalMethod method;
         // Lookup first in the metaclass as we are likely to find the method there
         final ModuleFields fields = Layouts.MODULE.getFields(metaClass);
         InternalMethod topMethod;
         if (noPrependedModulesProfile.profile(fields.getFirstModuleChain() == fields) &&
                 onMetaClassProfile.profile((topMethod = fields.getMethod(name)) != null) &&
-                !isRefinedProfile.profile(topMethod.isRefined())) {
+                !isRefinedProfile.profile(declarationContext != null && declarationContext.getRefinements() != null)) {
```

Comment:
@eregon could you help me with this, please?

I'm not 100% sure that it is right. Since there is no refined methods anymore I guess we should remove condition porfile and leave just condition itself: `declarationContext != null && declarationContext.getRefinements() != null`

Reply:
It looks fine, except that `DeclarationContext#getRefinements` is either a non-empty map, or `Collections.emptyMap()`, right?
Maybe we should have `DeclarationContext.NO_REFINEMENTS = Collections.emptyMap()` and then compare with that. That way we can also change the value for "no refinements" more easily if wanted.
Could also add a new `declarationContext.hasRefinements()` method, which would avoid that detail here.

`null` might be slightly better as a default than `Collections.emptyMap()` since then we don't have polymorphism on 2 different Map classes. But then we can't assume it's actually a Map anymore, so that seems inconvenient in some callers.

Reply:
We should keep the ConditionProfile though, as both cases are possible, and the JIT cannot always fold that condition.
We can rename though, maybe to `hasRefinementsProfile`?

Reply:
BTW, this is a small extra overhead, as before we'd just read a field of the `topMethod` and now we have a couple more indirections, but it seems acceptable as this is the fast-path uncached method lookup, which already does more expensive things like a method table lookup.

Reply:
> is either a non-empty map, or Collections.emptyMap(), right? ..

It was not strict, but I fixed it and used a new constant to refactor the code: https://github.com/oracle/truffleruby/pull/2033/commits/d6e6abf47dc4f3509cab215c7686868dac0ae56e